src/tensor_ops/cuda_kernels.rs

```diff
@@ -0,0 +1,172 @@
+    const PTX_SRC: &'static str;
+
+    /// Unique name for the kernel
+    const MODULE_NAME: &'static str;
+
+    /// Name of function in the .cu file
+    const FWD_FN_NAME: &'static str;
+
+    /// Name of function in the .cu file
+    const BWD_FN_NAME: &'static str;
+
+    const ALL_FN_NAMES: [&'static str; 2] = [Self::FWD_FN_NAME, Self::BWD_FN_NAME];
+}
+
+impl<K: BinaryOpCudaKernel> BinaryKernel<K, f32> for Cuda {
+    fn forward<S: Shape>(
+        &self,
+        _: K,
+        lhs: &Self::Storage<S, f32>,
+        rhs: &Self::Storage<S, f32>,
+    ) -> Result<Self::Storage<S, f32>, Self::Err> {
+        if !self.dev.has_func(K::MODULE_NAME, K::FWD_FN_NAME) {
+            self.dev
+                .load_ptx(K::PTX_SRC.into(), K::MODULE_NAME, &K::ALL_FN_NAMES)?;
+        }
+
+        let shape = lhs.shape;
+        let strides = lhs.shape.strides();
+        let numel = shape.num_elements();
+
+        let mut storage = self.dev.alloc_zeros_async::<f32>(numel)?;
+
+        let dims: CudaSlice<usize> = self.dev.take_async(shape.concrete().into())?;
+        let lhs_strides: CudaSlice<usize> = self.dev.take_async(lhs.strides.into())?;
+        let rhs_strides: CudaSlice<usize> = self.dev.take_async(rhs.strides.into())?;
+        let out_strides: CudaSlice<usize> = self.dev.take_async(strides.into())?;
+
+        let fwd_fn = self.dev.get_func(K::MODULE_NAME, K::FWD_FN_NAME).unwrap();
+        let cfg = LaunchConfig::for_num_elems(numel as u32);
+        let params = (
+            numel,             // const size_t numel,
+            S::NUM_DIMS,       // const size_t num_dims,
+            &dims,             // const size_t *dims,
+            lhs.data.as_ref(), // const float *lhs,
+            &lhs_strides,      // const size_t *lhs_strides,
+            rhs.data.as_ref(), // const float *rhs,
+            &rhs_strides,      // const size_t *rhs_strides,
+            &mut storage,      // float *out,
+            &out_strides,      // const size_t *out_strides
+        );
+        unsafe { fwd_fn.launch_async(cfg, params) }?;
+        Ok(CudaArray {
+            data: Arc::new(storage),
+            shape,
+            strides,
+        })
+    }
+
+    fn backward<S: Shape>(
+        &self,
+        _: K,
+        lhs: &Self::Storage<S, f32>,
+        grad_lhs: &mut Self::Storage<S, f32>,
+        rhs: &Self::Storage<S, f32>,
+        grad_rhs: &mut Self::Storage<S, f32>,
+        grad_out: &Self::Storage<S, f32>,
+    ) -> Result<(), Self::Err> {
+        let bwd_fn = self.dev.get_func(K::MODULE_NAME, K::BWD_FN_NAME).unwrap();
+        let numel = lhs.shape.num_elements();
+
+        let dims: CudaSlice<usize> = self.dev.take_async(lhs.shape.concrete().into())?;
+        let lhs_strides: CudaSlice<usize> = self.dev.take_async(lhs.strides.into())?;
+        let rhs_strides: CudaSlice<usize> = self.dev.take_async(rhs.strides.into())?;
+        let out_strides: CudaSlice<usize> = self.dev.take_async(grad_out.strides.into())?;
```

Comment:
These same values were also allocated in the forward call - a potential improvement for the future is pre-allocating them. Though these are only used in binary ops - if a tensor is only ever used in a unary op then it doesn't need to allocate these